File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/dialogue/config/PersonaVoiceProvider.java
package com.study.webflux.rag.infrastructure.dialogue.config;

import java.util.Map;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.stereotype.Service;

import com.study.webflux.rag.domain.dialogue.model.PersonaId;
import com.study.webflux.rag.domain.voice.model.AudioFormat;
import com.study.webflux.rag.domain.voice.model.Voice;
import com.study.webflux.rag.domain.voice.model.VoiceSettings;
import com.study.webflux.rag.domain.voice.model.VoiceStyle;
import com.study.webflux.rag.infrastructure.dialogue.config.properties.RagDialogueProperties;
import com.study.webflux.rag.infrastructure.dialogue.config.properties.RagDialogueProperties.PersonaVoiceConfig;

/**
 * 페르소나별 TTS Voice를 제공하는 서비스. application.yml의 personas 설정을 기반으로 각 페르소나에 맞는 Voice 객체를 반환합니다.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class PersonaVoiceProvider {

	private final RagDialogueProperties properties;
	private final Voice defaultVoice;

	/**
	 * 주어진 페르소나에 해당하는 Voice를 반환합니다. 설정이 없는 페르소나는 기본 Voice를 반환합니다.
	 *
	 * @param personaId
	 *            페르소나 식별자
	 * @return 페르소나에 맞는 Voice 객체
	 */
	public Voice getVoiceForPersona(PersonaId personaId) {
		if (personaId == null || PersonaId.DEFAULT.equals(personaId)) {
			return defaultVoice;
		}

		String personaKey = personaId.value();
		Map<String, PersonaVoiceConfig> personas = properties.getPersonas();

		if (personas.containsKey(personaKey)) {
			PersonaVoiceConfig config = personas.get(personaKey);
			return buildVoiceFromConfig(personaKey, config);
		}

		log.warn("페르소나 '{}'에 대한 보이스 설정이 없어 기본 보이스 사용", personaKey);
		return defaultVoice;
	}

	/**
	 * PersonaVoiceConfig를 Voice 도메인 객체로 변환합니다.
	 */
	private Voice buildVoiceFromConfig(String personaKey, PersonaVoiceConfig config) {
		var voiceSettingsConfig = config.getVoiceSettings();
		VoiceSettings settings = new VoiceSettings(voiceSettingsConfig.getPitchShift(),
			voiceSettingsConfig.getPitchVariance(), voiceSettingsConfig.getSpeed());

		return Voice.builder().id(config.getVoiceId()).name(personaKey).provider("supertone")
			.settings(settings).language(config.getLanguage())
			.style(VoiceStyle.fromString(config.getStyle()))
			.outputFormat(AudioFormat.fromString(properties.getSupertone().getOutputFormat()))
			.build();
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/dialogue/config/PersonaVoiceProvider.kt
package com.miyou.app.infrastructure.dialogue.config

import com.miyou.app.common.model.AudioFormat
import com.miyou.app.domain.voice.model.Voice
import com.miyou.app.domain.voice.model.VoiceSettings
import com.miyou.app.domain.voice.model.VoiceStyle
import com.miyou.app.domain.voice.port.VoiceSelectionPort
import com.miyou.app.infrastructure.dialogue.config.properties.RagDialogueProperties
import com.miyou.app.infrastructure.dialogue.config.properties.RagDialogueProperties.PersonaVoiceConfig
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.stereotype.Service

@Service
class PersonaVoiceProvider(
    private val properties: RagDialogueProperties,
    private val defaultVoice: Voice,
) : VoiceSelectionPort {
    private val log = KotlinLogging.logger {}

    override fun getVoiceForPersona(personaId: String): Voice {
        if (personaId.isBlank() || personaId == DEFAULT_PERSONA_KEY) {
            return defaultVoice
        }

        val personaKey = personaId
        val personas = properties.personas
        val config: PersonaVoiceConfig? = personas[personaKey]
        if (config != null) {
            return buildVoiceFromConfig(personaKey, config)
        }

        log.warn { "페르소나 '$personaKey'에 대한 설정을 찾지 못해 기본 음성 반환" }
        return defaultVoice
    }

    private fun buildVoiceFromConfig(
        personaKey: String,
        config: PersonaVoiceConfig,
    ): Voice {
        val voiceSettingsConfig = config.voiceSettings
        val settings =
            VoiceSettings(
                voiceSettingsConfig.pitchShift,
                voiceSettingsConfig.pitchVariance,
                voiceSettingsConfig.speed,
            )

        return Voice
            .builder()
            .id(config.voiceId)
            .name(personaKey)
            .provider("supertone")
            .settings(settings)
            .language(config.language)
            .style(VoiceStyle.fromString(config.style))
            .outputFormat(AudioFormat.fromString(properties.supertone.outputFormat))
            .build()
    }

    companion object {
        private const val DEFAULT_PERSONA_KEY = "default"
    }
}
